integrals: fixed bug which produced a wrong result when integrating derivatives

2torus opened the pull request:
Fixes issue #12899. The original logic was faulty in several ways:
a) treating a derivative with respect to a variable different
from integration symbol as if it was constant.
b) looking only at the last variable of the differentiation.
Fixed a test in test_manual. Added new tests in test_manual
This makes a mild assumption that integration and differentiation commute.

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

2torus committed:
integrals: fixed bug which produced a wrong result when integrating derivatives

Fixes issue #12899. The original logic was faulty in several ways:
a) treating a derivative with respect to a variable different
from integration symbol as if it was constant.
b) looking only at the last variable of the differentiation.
Fixed a test in test_manual. Added new tests in test_manual
This makes a mild assumption that integration and differentiation commute.

asmeurer commented:
Currently in master

```py
>>> diff(Integral(f(x, y), y), x)
Integral(Derivative(f(x, y), x), y)
```

So we already assume that. But I wonder if we should pick one, as otherwise it would just alternate on repeated calls to `doit`. Or maybe that's a good thing.

asmeurer commented on sympy/integrals/manualintegrate.py:
```diff
@@ -1153,7 +1163,17 @@ def eval_derivativerule(integrand, symbol):
+        # from the list of differentiating variables
+        diff_vars = tuple(reversed(integrand.args[1:]))
+        symbol_index = diff_vars.index(symbol)
+        diff_vars_symbol_removed = diff_vars[:symbol_index] + diff_vars[symbol_index + 1:]
```
You should be able to use `diff_vars.remove(symbol)` (if `diff_vars` is a list).

2torus replied:
It's a tuple at the moment but making it a list is more natural here.

asmeurer commented on sympy/integrals/manualintegrate.py:
```diff
@@ -1153,7 +1163,17 @@ def eval_derivativerule(integrand, symbol):
     else:
-        return sympy.Derivative(integrand.args[0], *integrand.args[1:-1])
+        # we are removing the last occurence of symbol
+        # from the list of differentiating variables
```
Differentiation variables are sorted. The logic here is not needed.

2torus replied:
Logic removed.

jksuom commented:
This looks good. Only some lines have grown out of bounds.

2torus committed:
integrals: fix bug when integrating derivative, small fixes

Some fixes after pull request review:
line lengths
logic and code change when removing differentiating variables.

2torus commented:
@jksuom I'll break up the long lines I introduced.

2torus commented:
@asmeurer Let me marinate on your remark regarding exchanging integral and derivatives, I might have been overzealous there. 
I fixed everything else in the last commit.

jksuom commented on sympy/integrals/tests/test_manual.py:
```diff
@@ -316,5 +317,11 @@ def test_issue_10847():
+
+
+
+def test_constant_independeint_of_symbol():
```
typo: independeint

2torus committed:
integrals: fix bug when integrating derivative, more conservative approach

Fix the original faulty logic but don't move the derivative w.r.t.
to other free symbols outside of the integral but just leave it as is.

2torus commented:
@asmeurer @jksuom Could you review the latest changes?
